### PR TITLE
Update debian/compat to version 13

### PR DESCRIPTION
**Pull request by janisozaur**
Compat version 13 is currently the recommended one.

An important change introduced in 10 was change of default to target parallel builds

https://github.com/Debian/debhelper/blob/5d1bb29841043d8e47ebbdd043e6cd086cad508e/debhelper.pod#compatibility-levels

### CODE REVIEW

File: scripts/release_ppa.sh
Comment (cameel) [inline scripts/release_ppa.sh:159]: We have one more script for building PPA packages. I think it should be updated too:
https://github.com/ethereum/solidity/blob/83002f3a0e6276c3e4056d2413bd3008e0a336cb/scripts/deps-ppa/static_z3.sh#L76

**Review by cameel — COMMENTED** [review]
Thanks for the PR! Looks like a good change.

It would be best to make sure that the script works without issues before we merge. We only ever run it manually and don't have a CI job that would check it so if it's broken, we'll only discover that when we'll be making a release. I guess, if that happens, we can work around by going back to 9 and looking into it after the release so maybe it's not such a big issue with this particular PR.

In any case, I looked through the linked changelog and didn't see anything obvious that we'd need to update to work with version 13 so it's probably fine. It's pretty detailed though so I could easily have missed something that would require changes on our side.

@ekpyron I see that `dput` has a `--check-only` option that builds and installs the package locally with all checks, without actually uploading it. Do you think it would be worthwhile to have a CI job that would run with this flag to verify these scripts?

File: scripts/release_ppa.sh
Comment (janisozaur) [reply scripts/release_ppa.sh:159]: done

**Comment by janisozaur** [comment]
In another package I've updated, I was able to confirm my changes with `debuild -b -uc -us` executed in `debian/`-enabled repository

**Comment by ekpyron** [comment]
Usually on any changes to this script we run it to create a prerelease on launchpad (we can do that at any time). A CI run for this seems a bit of an overkill, this script hardly ever changes :-).

**Comment by ekpyron** [comment]
https://manpages.ubuntu.com/manpages/focal/man7/debhelper.7.html#compatibility%20levels
would suggest that we may also need to bump the ``debhelper`` version in `debian/control` as a dependency, though?
``g++ (>= 5.0)`` could also have used a bump ages ago I guess :-).

**Comment by janisozaur** [comment]
I'll update those as well

**Comment by janisozaur** [comment]
I've updated the mentioned files. I've only updated gcc for solidity itself, I think z3 might support older versions, even though I wasn't able to verify that in their CI/docs immediately.

I've updated GCC to 9+, this is currently used in your CI setup.

**Comment by axic** [comment]
Btw the copyright section in the package is really outdated, the complete license is [here](https://github.com/ethereum/solidity/blob/develop/cmake/templates/license.h.in)

**Comment by janisozaur** [comment]
The license changes would fall outside of scope I intended for this PR. Can it be merged now?

File: scripts/deps-ppa/static_z3.sh
Comment (axic) [inline scripts/deps-ppa/static_z3.sh]: Why not use `>= 9.0` here too?

btw @ekpyron the [documentation](https://docs.soliditylang.org/en/v0.8.17/installing-solidity.html) still recommends gcc 8 as the minimum

Comment (janisozaur) [inline scripts/deps-ppa/static_z3.sh]:
```suggestion
               g++ (>= 9.0),
```

**Review by cameel — DISMISSED** [review]
It seems that all the review comments have been addressed a while ago so I'm approving. I'm going to leave merging it up to @ekpyron to give him a chance to have a final look (and maybe create a prerelease to test it).

**Comment by cameel** [comment]
Also, review fixes need to be squashed into the original commit before we merge.

**Comment by cameel** [comment]
I just tried to create a prerelease from this PR and unfortunately it fails for some reason:

```
dpkg-source: info: using source format '3.0 (quilt)'
dpkg-source: info: building solc using existing ./solc_0.8.18~develop-2022-12-06-51b4861e.orig.tar.gz
dpkg-source: error: cannot represent change to deps/downloads/fmt-8.0.1.tar.gz: binary file contents changed
dpkg-source: error: add deps/downloads/fmt-8.0.1.tar.gz in debian/source/include-binaries if you want to store the modified binary in the debian tarball
dpkg-source: error: unrepresentable changes to source
dpkg-buildpackage: error: dpkg-source -b . subprocess returned exit status 1
debuild: fatal error at line 1182:
dpkg-buildpackage -us -uc -ui -S -d -sa failed
```

I tried running on `develop` and there it works. On top of this PR it failed twice. We'll need to figure out what's happening. I'm leaving it for now because we have a release coming and need to focus on getting PRs ready for that. This one unfortunately won't make it.

### Repro
Here's how I was running it:

```bash
docker run -it --rm --volume /tmp/gnupg:/root/.gnupg debian
```

```bash
branch=debian-compat-13-ppa

cd
apt update
apt install --assume-yes --no-install-recommends git wget devscripts debhelper dput

# My key has Unicode characters and makes gpg fail with the default POSIX locale
apt install --assume-yes --no-install-recommends locales
sed -i 's/^# *\(en_US\.UTF-8\)/\1/' /etc/locale.gen
locale-gen
export LC_ALL=en_US.UTF-8

git clone https://github.com/ethereum/solidity --depth 1 --branch "$branch"
cd solidity/
git branch --show-current

cat << EOF > /etc/dput.cf 
[ethereum-dev]
fqdn            = ppa.launchpad.net
method          = ftp
incoming        = ~ethereum/ethereum-dev
login           = anonymous

[ethereum]
fqdn            = ppa.launchpad.net
method          = ftp
incoming        = ~ethereum/ethereum
login           = anonymous

[ethereum-static]
fqdn            = ppa.launchpad.net
method          = ftp
incoming        = ~ethereum/ethereum-static
login           = anonymous
EOF

cat << EOF > .release_ppa_auth
export LAUNCHPAD_EMAIL=<my mail>
export LAUNCHPAD_KEYID=<my key id>
EOF

scripts/release_ppa.sh "$branch"
```

BTW, I had to create a branch called `debian-compat-13-ppa` pointing at this PR because `release_ppa.sh` expects a branch from our repo (not a fork). We can delete it when the PR is merged.

**Comment by janisozaur** [comment]
The error message suggests it's this issue: https://manpages.debian.org/stretch/dpkg-dev/dpkg-source.1.en.html#cannot_represent_change_to_file but do you modify the file in question? Is it referenced anywhere in the debian package specs? I can't find such references anywhere.

**Comment by ekpyron** [comment]
Fun fact: we just figured out that apparently multiple fetches of https://github.com/fmtlib/fmt/archive/8.0.1.tar.gz yield archives with different hashes, albeit with identical contents, that seems to be causing the issue - probably unrelated to the change in this PR.

**Comment by cameel** [comment]
Looks like it's not related to the PR after all.

The file is indeed different but this seems to be a weird issue on Github. The source link suddenly started non-deterministically returning different files. If you run this multiple times:
```bash
curl https://github.com/fmtlib/fmt/archive/8.0.1.tar.gz --remote-name --location --silent && sha256sum 8.0.1.tar.gz
```
you'll eventually get different hashes. Either `b06ca3130158c625848f3fb7418f235155a4d389b2abc3a6245fb01cb0eb1e01` or `a811a7d8b49694514b49059ae390f2417d6c53b6ec3f1e83a3a65619ed37375f`. Both are valid archives with the same exact content (diffs binary equal) but the archives themselves are different (one is ~2 kB bigger).

It appears that the source archive for version 9.1.0 has the same problem. Might be a more widespread github issue. Our CI builds are now regularly failing due to this.

This is breaking our release scripts and we have a release planned for tomorrow so it's a big issue we're trying to work around now.

**Comment by janisozaur** [comment]
I think you may already be involved, but just in case: https://github.blog/changelog/2023-01-30-git-archive-checksums-may-change/

**Comment by cameel** [comment]
Now that we're done with releasing solc, I tried doing a pre-release PPA build again and it succeeded as expected. The Z3 build seems to work as well locally, though I could not really test it fully because we already have the latest version built and Launchpad rejects it for that reason. At least does not seem like it fails due to any other problem.

I think we can merge it once CI finishes.

**Review by cameel — APPROVED** [review]

**Comment by ekpyron** [comment]
> Now that we're done with releasing solc, I tried doing a pre-release PPA build again and it succeeded as expected. The Z3 build seems to work as well locally, though I could not really test it fully because we already have the latest version built and Launchpad rejects it for that reason. At least does not seem like it fails due to any other problem.
> 
> I think we can merge it once CI finishes.

From the looks of it, the prerelease was created from the old script without this PR, though - and apparently ubuntu focal actually is incompatible with the change (see https://launchpadlibrarian.net/652849231/buildlog_ubuntu-focal-amd64.solc_1%3A0.8.19-0ubuntu1~focal_BUILDING.txt.gz) - so we'll need to revert this PR, unfortunately.

**Pull request merged**
